algorithm/pge-eda.metta

```diff
@@ -0,0 +1,564 @@
+            ($index (% $i L)))
+
+        (selectByIndex $g $index)))
+
+;; updateIndex -- increments the codon-index value by 1 when called
+;;              $tSpace -- genotype space
+;;              $random -- random number to prevent metta from caching results
+;;                      $i -- old index
+;;                      $next -- new index (+ 1 $i)
+
+(= (updateIndex $tSpace $random)
+    (let*
+        (
+            ($i (match $tSpace (codon-index $i) $i))
+            (() (remove-atom $tSpace (codon-index $i)))
+            ($next (+ 1 $i))
+        )
+        (add-atom $tSpace (codon-index $next))))
+
+;; ::::::::::::: GENERATE PHENOTYPE :::::::::::::
+;; genPhen -- generates phenotype starting from start symbol and using the genotype information to map productio rules
+;;              $gSpace -- grammar space
+;;              $tSpace -- genotype space
+;;              $nts -- set of non-terminals
+;;              $start -- the start symbol
+;;              $maxIter -- maximum iteration length
+;;              $random -- random number to prevent metta from caching results
+;;                      $mapped -- a terminal symbol after having been transformed from a non-terminal using the grammar 
+;;                      $codon -- a probablity value gotten from &tSpace 
+;;                      $r -- variable to store recursive call to the function genPhen
+;;                      $codon -- a probablity value gotten from &tSpace 
+;;                      $nextRule -- production rule picked from the set of available expansions of a non-terminal based on probability values
+;;                      $codonIndex -- the updated condon index value
+
+(= (genPhen $gSpace $cSpace $tSpace $nts $start $maxIter $random)
+    (if (and (isLiteral $start) (not (isMember $start $nts)))
+        $start
+        (if (== (get-metatype $start) Expression)
+
+            (let* 
+                (
+                    (($f $t) (decons $start))
+                    ($mapped (genPhen $gSpace $cSpace $tSpace $nts $f $maxIter ((py-atom random.random))))
+                    )
+
+                (if (== $t ())
+                    (cons-atom $mapped $t)
+                    (let $r (genPhen $gSpace $cSpace $tSpace $nts $t $maxIter ((py-atom random.random)))
+                        (cons-atom $mapped $r) )))                  
+
+            (let*
+                (   
+                    ($codon (getCodon $tSpace ((py-atom random.random))))
+                    ($nextRule (getRule $start $gSpace $cSpace $codon 0 0 (py-atom random.random)))
+                
+                    (() (println! $nextRule))
+                    (() (updateIndex $tSpace ((py-atom random.random))))
+                    ($codonIndex (match $tSpace (codon-index $i) $i)))
+
+                (if (< $codonIndex $maxIter)
+                    (genPhen $gSpace $cSpace $tSpace $nts $nextRule $maxIter ((py-dot np random.uniform)0 1))
+
+                    X )))))
```

Comment:
Just wanted to make sure I got it right. Are you returning `X` here because the boolean expression won't just stop right?